File: springboot-jwt/src/main/java/com/grepp/spring/infra/auth/token/JwtProvider.java
package com.grepp.spring.infra.auth.token;

import com.grepp.spring.app.model.auth.token.RefreshTokenRepository;
import com.grepp.spring.app.model.auth.domain.Principal;
import com.grepp.spring.app.model.auth.token.dto.AccessTokenDto;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.ExpiredJwtException;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.MalformedJwtException;
import io.jsonwebtoken.UnsupportedJwtException;
import io.jsonwebtoken.security.Keys;
import io.jsonwebtoken.security.SecurityException;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Base64;
import java.util.Date;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;
import javax.crypto.SecretKey;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
@Slf4j
public class JwtProvider {

    private RefreshTokenRepository refreshTokenRepository;
    private final AuthenticationManagerBuilder authenticationManagerBuilder;

    @Value("${jwt.secrete}")
    private String key;

    @Getter
    @Value("${jwt.access-expiration}")
    private long atExpiration;

    @Getter
    @Value("${jwt.refresh-expiration}")
    private long rtExpiration;

    private SecretKey secretKey;

    public SecretKey getSecretKey(){
        if(secretKey == null){
            String base64Key = Base64.getEncoder().encodeToString(key.getBytes());
            secretKey = Keys.hmacShaKeyFor(base64Key.getBytes(StandardCharsets.UTF_8));
        }
        return secretKey;
    }

    public AccessTokenDto generateAccessToken(Authentication authentication){
        String authorities = authentication.getAuthorities().stream()
            .map(GrantedAuthority::getAuthority)
            .collect(Collectors.joining(","));

        String id = UUID.randomUUID().toString();
        long now = new Date().getTime();
        Date atExpiresIn = new Date(now + atExpiration);
        String accessToken = Jwts.builder()
            .subject(authentication.getName())
            .id(id)
            .claim("auth",authorities)
            .expiration(atExpiresIn)
            .signWith(getSecretKey())
            .compact();

        return AccessTokenDto.builder()
            .id(id)
            .token(accessToken)
            .expiresIn(atExpiration)
            .build();
    }

    public Authentication genreateAuthentication(String accessToken){
        Claims claims = parseClaim(accessToken);
        List<? extends GrantedAuthority> authorities =
            Arrays.stream(claims.get("auth").toString().split(","))
                .map(SimpleGrantedAuthority::new)
                .toList();

        Principal principal = new Principal(claims.getSubject(),"", authorities);
        return new UsernamePasswordAuthenticationToken(principal, "", authorities);
    }

    public Claims parseClaim(String accessToken) {
        try{
            return Jwts.parser().verifyWith(getSecretKey()).build()
                .parseSignedClaims(accessToken).getPayload();
        }catch (ExpiredJwtException ex){
            return ex.getClaims();
        }
    }


    public boolean validateToken(String requestAccessToken) {
        try{
            Jwts.parser().verifyWith(getSecretKey()).build().parse(requestAccessToken);
            return true;
        }catch(SecurityException | MalformedJwtException | UnsupportedJwtException | IllegalArgumentException e){
            log.error(e.getMessage(), e);
        }
        return false;
    }
}

File: grepp-gateway/src/main/kotlin/com/grepp/spring/token/util/JwtProvider.kt
package com.grepp.spring.token.util

import com.grepp.spring.token.code.TokenType
import com.grepp.spring.token.dto.AccessTokenDto
import com.grepp.spring.token.repository.RefreshTokenRepository
import io.jsonwebtoken.*
import io.jsonwebtoken.security.Keys
import io.jsonwebtoken.security.SecurityException
import jakarta.servlet.http.Cookie
import jakarta.servlet.http.HttpServletRequest
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.core.Authentication
import org.springframework.stereotype.Service
import java.nio.charset.StandardCharsets
import java.util.*
import javax.crypto.SecretKey

@Service
class JwtProvider(
    val refreshTokenRepository: RefreshTokenRepository
) {
    private val log = LoggerFactory.getLogger(javaClass)

    @Value("\${jwt.secrete}")
    val key: String? = null

    @Value("\${jwt.access-expiration}")
    val atExpiration: Long = 0

    @Value("\${jwt.refresh-expiration}")
    val rtExpiration: Long = 0

    var secretKey: SecretKey? = null
        get() {
            if (field == null) {
                val base64Key = Base64.getEncoder().encodeToString(key!!.toByteArray())
                field = Keys.hmacShaKeyFor(base64Key.toByteArray(StandardCharsets.UTF_8))
            }
            return field
        }

    fun generateAccessToken(authentication: Authentication): AccessTokenDto {
        return generateAccessToken(authentication.name, authentication.authorities.joinToString(","))
    }

    fun generateAccessToken(username: String, roles: String): AccessTokenDto {
        val id = UUID.randomUUID().toString()
        val now = Date().time
        val atExpiresIn = Date(now + atExpiration)
        val accessToken = Jwts.builder()
            .subject(username)
            .id(id)
            .claim("roles", roles)
            .expiration(atExpiresIn)
            .signWith(secretKey)
            .compact()

        return AccessTokenDto(
            id=id,
            token = accessToken,
            expiresIn = atExpiration
        )
    }

    fun parseClaim(accessToken: String?): Claims? {
        return try {
            Jwts.parser().verifyWith(secretKey).build()
                .parseSignedClaims(accessToken).payload
        } catch (ex: ExpiredJwtException) {
            ex.claims
        } catch (ex: Exception){
            log.info(ex.message, ex)
            return null
        }
    }

    fun validateToken(requestAccessToken: String?): Boolean {
        try {
            Jwts.parser().verifyWith(secretKey).build().parse(requestAccessToken)
            return true
        } catch (e: SecurityException) {
            log.error(e.message, e)
        } catch (e: MalformedJwtException) {
            log.error(e.message, e)
        } catch (e: UnsupportedJwtException) {
            log.error(e.message, e)
        } catch (e: IllegalArgumentException) {
            log.error(e.message, e)
        }
        return false
    }

    fun resolveToken(request: HttpServletRequest, tokenType: TokenType): String? {
        val headerToken = request.getHeader("Authorization")
        if (headerToken != null && headerToken.startsWith("Bearer")) {
            return headerToken.substring(7)
        }

        val cookies = request.cookies ?: return null

        return Arrays.stream(cookies)
            .filter { e: Cookie -> e.name == tokenType.name }
            .map { obj: Cookie -> obj.value }.findFirst()
            .orElse(null)
    }
}